### DISCUSSION TITLE
Merkle Tree Implementation over current multicall disbursement

### DISCUSSION THREAD

**Discussion by raghav-rumsan** [rahataid/rahat-platform#198] [Ideas]
## 🛠 Context

In systems where financial aid, tokens, or resources are disbursed to multiple beneficiaries, choosing the right mechanism for tracking allocations and verifying disbursements is critical. We explore the application of **Merkle Trees** for managing and validating disbursement data while contrasting it with alternative methods like **multicall** and **Merkle Forests**.

## 💡 Concept Overview

Imagine a blockchain-based disbursement system where:
1. Beneficiaries are allocated varying amounts of resources upon a trigger event (e.g., disaster aid or financial incentives).
2. Disbursements occur in **stages** (e.g., 10% or fixed amounts over time).
3. Validation is essential to ensure transparency, prevent tampering, and enable scalability.

### Why Merkle Trees?
Merkle Trees provide an efficient way to:
- Represent beneficiary data in a compact, tamper-proof structure.
- Validate individual claims without exposing the entire dataset.
- Ensure cryptographic integrity by storing only the Merkle Root on-chain.

---

### Multicall Advantages:
- Simplicity: Executes multiple operations in a single transaction.
- Transparency: All data updates occur on-chain.

### Challenges:
1. **Cost:** Every beneficiary update requires on-chain computation and storage, leading to high gas fees.
2. **Scalability:** Multicall becomes inefficient as the number of beneficiaries grows.
3. **State Exposure:** All updates are visible, which might compromise privacy.

---

## 🌳 Merkle Tree Over Merkle Forest: A Decision Framework

**Merkle Tree Approach**:
- **Use Case:** Best suited for systems with a single root of authority (e.g., one-time disbursements or fixed allocation tracking).
- **Efficiency:** Simple to implement and maintain when the beneficiary pool size is manageable.

**Merkle Forest Approach**:
- **Use Case:** Ideal for multi-step, time-bound disbursements where stages or categories of beneficiaries need distinct roots.
- **Efficiency:** Each category (or stage) has its own tree, avoiding frequent recalculations of the entire dataset.
- **Scalability:** Supports hierarchical structures and allows parallel updates without compromising the integrity of other trees.

---

## 🚀 Why Choose Merkle Trees for Disbursements?

1. **Efficiency in Verification:** Beneficiaries can validate their claims off-chain using proofs, reducing the load on the blockchain.
2. **Cost-Effectiveness:** Only the Merkle Root needs to be updated on-chain, minimizing gas costs.
3. **Privacy:** Individual proofs reveal minimal information about the dataset.
4. **Flexibility:** Easy to handle partial claims or staged disbursements.

---

## 🌟 Open Questions for Discussion

1. **Merkle Tree vs. Forest:** For staged disbursements, when is a single tree sufficient, and when should we opt for multiple roots (forest)?
2. **Handling Partial Claims:** How can we ensure consistency if a beneficiary claims only part of their allocation? What are the edge cases in updating the tree?
3. **Alternative Methods:** Could approaches like multicall or hybrid solutions (e.g., off-chain state management with on-chain validation) be more suitable for specific scenarios?
4. **Trigger Mechanisms:** How can real-world events (e.g., climate data or economic indicators) best integrate with a Merkle-based disbursement framework?

---

## 🎯 Call to Action

We invite contributors, developers, and blockchain enthusiasts to share insights, propose alternative methods, or suggest enhancements to this model. Let’s collaboratively explore how to create a transparent, scalable, and cost-effective disbursement system for blockchain-based applications. 

Join the discussion below! 🚀

Would love to get your ideas and concepts on this @sushant-rumsan  @manjik-rumsan

**Comment by raghav-rumsan** [comment]
https://github.com/rahataid/rahat-platform/tree/merkle-tree

https://github.com/rahataid/rahat-platform/blob/da593ea19bc1a8963eda1a20fc5ae59ec1fd590c/apps/contracts/src/contracts/MerkleDisbursement.sol

https://github.com/argahv/merkle

**Comment by Fils7** [comment]
Hi @raghav-rumsan .Thanks for opening this discussion—these are critical design questions for scalable, transparent, and secure disbursement systems. I’d like to share some thoughts and suggestions on the points raised:

---------------------------------------------------------------

## Merkle Tree vs. Merkle Forest
I agree that a single Merkle tree is well-suited for one-off or fixed disbursements, where all allocations are set at once. For staged or category-based distributions (e.g., periodic aid), a Merkle forest—where each stage or category has its own root—offers better scalability and flexibility. This avoids the need to recalculate the entire tree for every update and allows parallel management of distinct disbursement rounds.

## Handling Partial Claims
Partial claims introduce some complexity. The main approaches I see are:

### On-chain tracking:
Store a mapping of claimed amounts per beneficiary (in [MerkleDisbursement.sol](https://github.com/rahataid/rahat-platform/blob/da593ea19bc1a8963eda1a20fc5ae59ec1fd590c/apps/contracts/src/contracts/MerkleDisbursement.sol)). When a claim is made, verify that the sum of all claims does not exceed the original allocation (proven via Merkle proof).
This method is straightforward and gas-efficient, but does require a small amount of per-user storage.

### Off-chain tree updates:
After each partial claim, update the user’s leaf off-chain, recalculate the Merkle root, and update it on-chain.
This keeps all state in the tree, but is costly and can create concurrency issues if multiple claims happen in quick succession.

### ZK nullifiers:
For privacy and double-claim prevention, you can use a nullifier hash mechanism, as seen in Semaphore. Each claim generates a unique nullifier, and the contract checks that it hasn’t been used before. For partial claims, this can be extended by encoding (user, claimed amount, nonce) in the nullifier, but the ZK circuit and contract logic become more complex.

```solidity
mapping(address => uint256) public claimedAmounts;

function claim(uint256 totalAllocated, uint256 claimAmount, bytes32[] calldata proof) external {
    bytes32 leaf = keccak256(abi.encodePacked(msg.sender, totalAllocated));
    require(MerkleProof.verify(proof, merkleRoot, leaf), "Invalid proof");
    require(claimedAmounts[msg.sender] + claimAmount <= totalAllocated, "Over-claiming");
    claimedAmounts[msg.sender] += claimAmount;
    // Transfer claimAmount to user...
}
```

Also check this documentation from Metaplex that can be of interest: [How to Create a Token Claimer Smart Contract leveraging Merkle Tree](https://developers.metaplex.com/token-metadata/guides/anchor/token-claimer-smart-contract)


## Alternative Methods
A hybrid approach could be valuable—using multicall for small, high-trust batches and Merkle proofs for larger privacy-sensitive, or staged distributions. This flexibility lets admins choose the best mechanism per scenario.

## Trigger Mechanisms & Oracle Security
Oracles are powerful for automating disbursements based on real-world events, but they are also a frequent attack vector (oracle manipulation and fake data are common in DeFi exploits). To mitigate this:

1. Use decentralized oracles that aggregate multiple sources (e.g., Chainlink).

2. Implement sanity checks and thresholds in the contract to reject extreme or suspicious data.

3. For critical disbursements, consider requiring manual or multi-sig approval in addition to oracle triggers.

### Open to Collaboration
I’d be happy to help prototype a ZK/nullifier-based partial claim system, or work on the smart contracts if you see any think we could add to the current development. Looking forward to feedback and further collaboration!
